### PR TITLE
[Performance] Introducing Prefix-Cached Chunked Prefill with flash-attn backend and 10% throughput gained under prompt <1K

### PR DESCRIPTION
**Pull request by Juelianqvq**
Fix illegal memory access Error & Misaligned result under high concurrency in https://github.com/vllm-project/vllm/pull/6144 
The root cause of this problem maybe the overlook of the ``token_chunk_size`` boundary check, leading to misaligned indices calculation.
The reviewers have made a mistake in that PR since offline consistency or single request 1-by-1 can be passed relatively easily but it can be wrong in high concurrency situations (batched inference + scheduler behaviour), even the prefix-caching length order matters. I use Locust to reproduce the problem.

**[Performance Related]**
having tested with following two extreme scenarios(vs Prefix Caching ONLY):

**Llama-2-13B:** 
500-600 input tokens(300 cached) + 50 output tokens  
reaching +15% throughput gained under 10-15 concurrency

**Qwen2-72B-AWQ:** 
2k-6k input tokens(1k cached) + 512 output tokens 
reaching +3.5% throughput gained under 10 concurrency
(Maybe Disaggregated is the better choice)

Since my GPU is not good enough, welcome to use this PR and share your results.

**[Deficiency and Help Wanted]**
1.Don't know why cached length (means value of len(computed blocks)) can be modified when a sequence is being split into different rounds of chunks. So a little dirty code is attached as hack here (see the comments).
2.Compatibility has not been verified in a hurry, have a try and any feedback is welcome

Hope to find out the answer together @sighingnow 

cc  @rkooo567 @comaniac @Yard1 @zhuohan123 @WoosukKwon 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by comaniac** [comment]
Thanks for the PR. Can you elaborate why the illegal memory access is related to prefix caching + chunked prefill? At the current main we cannot enable them together so I'm a bit confused about why this would cause a problem. Meanwhile #6144 is also enabling prefix caching and chunked prefill. It would be good to consolidate the changes.

**Comment by comaniac** [comment]
(Please ignore my previous comment as I was referring to the wrong PR)

Thanks for the investigation. I'll review the PR tomorrow.

**Comment by robertgshaw2-neuralmagic** [comment]
Awesome work to run this down!

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -365,20 +368,36 @@ def _compute_for_prefix_cache_hit(
                             and self.sliding_window is None
                             and inter_data.is_prompt)
         inter_data.prefix_cache_hit = prefix_cache_hit
-        if self.chunked_prefill_enabled and prefix_cache_hit:
-            raise RuntimeError(
-                "chunked prefill cannot be used with prefix caching now.")
+        # if self.chunked_prefill_enabled and prefix_cache_hit:
```
Comment (rkooo567) [inline vllm/worker/model_runner.py:371]: remove it?

**Review by rkooo567 — COMMENTED** [review]
I will take a more deeper look, but does it also work with block manager v2? we will enable that by default soon I think

**Comment by comaniac** [comment]
> > I will take a more deeper look, but does it also work with block manager v2? we will enable that by default soon I think
> 
> Does v2's performance exceed v1? If so I'll dig into it in my spare time.

v2 has the similar performance as v1, but v2 supports more features such as lookahead slot allocation which is required for speculative decoding (and other ongoing features). Since our ultimate goal is deprecating v1, supporting v2 is necessary.

File: vllm/worker/model_runner.py [hunk 2]
```diff
@@ -365,20 +368,36 @@ def _compute_for_prefix_cache_hit(
...
+                inter_data.input_positions[seq_idx] = inter_data.input_positions[
+                    seq_idx][context_len:]
+            else:
+                if inter_data.cached_len > context_len:
+                    delta_len = min(inter_data.cached_len - context_len, seq_group_metadata.token_chunk_size - 1)
+                    context_len += delta_len
+                    inter_data.input_tokens[seq_idx] = inter_data.input_tokens[
+                        seq_idx][delta_len:]
+                    inter_data.input_positions[seq_idx] = inter_data.input_positions[
+                        seq_idx][delta_len:]
```
Comment (sighingnow) [inline vllm/worker/model_runner.py:399]: Hi @Juelianqvq, I have read the diff and if I understand it correctly, the key different between this PR with #6144 is: leaving at least 1 token for prefill for each sequence. I could add such logic into #6144.

**Pull request closed without merging**